### PR TITLE
Add unit tests for the Notices class and fix bugs

### PR DESCRIPTION
**Pull request by willmot**
This PR fixes up a bunch of issues in the Notices class and adds full unit test coverage. Issues resolved:

 * `(array) get_option` returned an array with a single empty element rather than an empty array.
 * Setting a persistent notice duplicated it as it was then in `$this->notices` and the database.
 * Calling `get_notices` with a context which didn't exist returned all notices rather than none.
 * Calling `set_notices` multiple times with same context overwrote all but the last messages.
 * It was possible to set broken messages with empty context or message.
 * `update_option` was needlessly called when setting a message that had already previously been set.

It also completes the docblocks and brings the Singleton pattern inline with Path.

- [x] Persistent notices aren't dismissible.

### CODE REVIEW

**Comment by willmot** [comment]
Would be good to catch the Scrutiniser issue before merging: https://scrutinizer-ci.com/g/humanmade/backupwordpress/inspections/fb6bf204-3ded-42f2-8c0e-21b8cf8a3b01/issues/files/classes/class-notices.php?status=new&orderField=path&order=asc&honorSelectedPaths=0

File: classes/class-notices.php [hunk 1]
```diff
@@ -76,19 +128,33 @@ public function get_notices( $context = '' ) {
...
+	 *
+	 * @return array The array of notices
 	 */
-	public function clear_all_notices() {
+	private function get_persistent_notices() {
+		$notices = get_option( 'hmbkp_notices' );
+		if ( ! empty( $notices ) ) {
+			return $notices;
+		}
+		return array();
```
Comment (pdewouters) [inline classes/class-notices.php]: How about `return ! empty( $notices ) ? $notices : array();`

Comment (willmot) [inline classes/class-notices.php]: Yeah that would be neater 👍 doing now.

File: classes/class-notices.php [hunk 2]
```diff
@@ -65,8 +117,8 @@ public function get_notices( $context = '' ) {
 
 		if ( $notices ) {
```
Comment (pdewouters) [inline classes/class-notices.php]: Should it be `if ( ! empty( $notices ) )` ?

**Comment by pdewouters** [comment]
is it good to review now?

**Comment by willmot** [comment]
@pdewouters yep!

**Pull request merged**